extensions/interactions/RatioExpressionInput/directives/ratio-expression-input-validation.service.ts

```diff
@@ -0,0 +1,200 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Validator service for the RatioExpressionInput interaction.
+ */
+
+import { Injectable } from '@angular/core';
+import { downgradeInjectable } from '@angular/upgrade/static';
+
+import { AnswerGroup } from
+  'domain/exploration/AnswerGroupObjectFactory';
+import { Warning, baseInteractionValidationService } from
+  'interactions/base-interaction-validation.service';
+import { RatioExpressionInputCustomizationArgs } from
+  'extensions/interactions/customization-args-defs';
+import { RatioObjectFactory, Ratio } from
+  'domain/objects/RatioObjectFactory';
+import { RatioExpressionInputRulesService } from
+  './ratio-expression-input-rules.service';
+import { Outcome } from
+  'domain/exploration/OutcomeObjectFactory';
+import { AppConstants } from 'app.constants';
+
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RatioExpressionInputValidationService {
+  constructor(
+    private rof: RatioObjectFactory,
+    private baseInteractionValidationServiceInstance:
+      baseInteractionValidationService) {}
+
+  getCustomizationArgsWarnings(
+      customizationArgs: RatioExpressionInputCustomizationArgs): Warning[] {
+    var isInt = function(n) {
+      return angular.isNumber(n) && n % 1 === 0;
+    };
+    var minimumNumberOfTerms =
+      customizationArgs.numberOfTerms.value;
+    if (minimumNumberOfTerms === undefined) {
+      return [
+        {
+          type: AppConstants.WARNING_TYPES.ERROR,
+          message: (
+            'Number of terms cannot be empty.')
+        }
+      ];
+    } else if (!isInt(minimumNumberOfTerms)) {
+      return [
+        {
+          type: AppConstants.WARNING_TYPES.ERROR,
+          message: (
+            'Number of terms must be integral.')
+        }
+      ];
+    } else {
+      return [];
+    }
+  }
+
+  getAllWarnings(
+      stateName: string,
+      customizationArgs: RatioExpressionInputCustomizationArgs,
+      answerGroups: AnswerGroup[], defaultOutcome: Outcome): Warning[] {
+    let warningsList = [];
+    let ratioRulesService = (
+      new RatioExpressionInputRulesService(this.rof));
+    var minimumNumberOfTerms =
+      customizationArgs.numberOfTerms.value;
+    warningsList = warningsList.concat(
+      this.getCustomizationArgsWarnings(customizationArgs));
+
+    warningsList = warningsList.concat(
+      this.baseInteractionValidationServiceInstance.getAllOutcomeWarnings(
+        answerGroups, defaultOutcome, stateName));
+
+    // This validations ensures that there are no redundant rules present in the
+    // answer groups.
+    // An Equals rule will make all of the following rules with a
+    // matching input, invalid.
+    // A HasNumberOfTermsEqualTo rule will make the following rules of the
+    // same rule type and a matching input, invalid.
```

Comment:
```suggestion
    // The following validations ensure that there are no redundant rules present in
    // the answer groups. In particular, an Equals rule will make all of the following rules
    // with a matching input invalid. A HasNumberOfTermsEqualTo rule will make the
    // following rules of the same rule type and a matching input invalid.
```

Suggested rewording--just some grammar nits. Also note how the documentation was changed to flow like a paragraph vs. being close-by single-line doc comments.

Reply:
done

Reply:
done